comictaggerlib/plugin_update_manager.py

```diff
@@ -0,0 +1,332 @@
+            if not name:
+                try:
+                    name = response.headers["content-disposition"].split("=")[1]
+                except Exception:
+                    name = url.rsplit("/", 1)[1]
+                finally:
+                    if not (name.endswith(".whl") or name.endswith(".zip")):
+                        logger.error("Ignoring download, unexpected or unknown file extension: %s", name)
+                        return False, Path()
+
+            filepath = self.plugin_dir.joinpath(name)
+
+            try:
+                with open(filepath, mode="wb") as file:
+                    file.write(response.content)
+                    return True, filepath
+            except Exception as e:
+                logger.exception("Failed to save plugin download file: %s. Error: %s", filepath, e)
+
+        return False, Path()
+
+    def _remove_old_plugins(self) -> None:
+        # Update local plugins first
+        self._find_local_plugins()
+
+        for key, item in self.plugin_files.items():
+            if len(item) > 1:
+                for i, (version, filename) in enumerate(item):
+                    # First filename should be latest version to keep
+                    if i > 0:
+                        self._remove_plugin(self.plugin_dir.joinpath(filename))
+
+    def _remove_plugin(self, file: Path) -> None:
+        filepath = self.plugin_dir.joinpath(file)
+        # Don't want to be nuking dirs by mistake
+        if filepath.is_file():
+            try:
+                os.remove(filepath)
+                logger.info(f"Removed old plugin file: {filepath}")
+            except Exception as e:
+                logger.exception(f"Failed to remove old plugin file: {filepath}. Error: {e}")
+
+    def _manifest_request(self, url: str) -> Any:
+        # if there is a 500 error, try a few more times before giving up
+        limit_counter = 0
+
+        for tries in range(1, 5):
+            try:
+                resp = requests.get(url, headers={"user-agent": "comictagger"}, timeout=10)
+                if resp.status_code == 200:
+                    return resp.text
+                elif resp.status_code == 500:
+                    logger.debug(f"Try #{tries}: ")
+                    time.sleep(1)
+                    logger.debug(str(resp.status_code))
+
+                elif resp.status_code in (requests.status_codes.codes.TOO_MANY_REQUESTS):
+                    logger.info(f"{url} rate limit encountered. Waiting for 10 seconds\n")
+                    time.sleep(10)
+                    limit_counter += 1
+                    if limit_counter > 3:
+                        # Tried 3 times, inform user to check CV website.
+                        logger.error(f"{url} rate limit error. Exceeded 3 retires.")
+                else:
+                    break
+
+            except requests.exceptions.Timeout:
+                logger.debug(f"Connection to {url} timed out.")
```

Comment:
Go with the simplest code possible and check whatever assumption is made, then let the user know what failed, they can retry themselves if they want to.